vng/vector/primitive.go

```diff
@@ -41,15 +87,88 @@ func (p *PrimitiveWriter) Flush(eof bool) error {
 	return err
 }
 
+func (p *PrimitiveWriter) makeDictVector() []byte {
+	dict := p.makeDict()
+	pos := make(map[string]byte)
+	for off, entry := range dict {
+		if bytes := entry.Value.Bytes(); bytes != nil {
+			pos[string(bytes)] = byte(off)
+		}
+	}
+	out := make([]byte, 0, p.count)
+	for it := p.bytes.Iter(); !it.Done(); {
+		bytes := it.Next()
+		if bytes == nil {
+			// null is always the first dict entry if it exists
+			out = append(out, 0)
+			continue
+		}
+		off, ok := pos[string(bytes)]
+		if !ok {
+			panic("bad dict entry") //XXX
+		}
+		out = append(out, off)
+	}
+	return out
+}
+
 func (p *PrimitiveWriter) Segmap() []Segment {
 	return p.segments
 }
 
+func (p *PrimitiveWriter) Const() *Const {
+	if len(p.dict)+p.hasNull != 1 {
+		return nil
+	}
+	var bytes zcode.Bytes
+	if len(p.dict) == 1 {
+		for b := range p.dict {
+			bytes = []byte(b)
+		}
+	}
+	return &Const{
+		Value: zed.NewValue(p.typ, bytes),
+		Count: p.count,
+	}
+}
+
 func (p *PrimitiveWriter) Metadata() Metadata {
+	var dict []DictEntry
+	if p.dict != nil {
+		if cnt := len(p.dict) + p.hasNull; cnt != 0 {
+			if cnt == 1 {
+				return p.Const()
+			}
+			dict = p.makeDict()
+		}
+	}
 	return &Primitive{
 		Typ:    p.typ,
 		Segmap: p.segments,
+		Dict:   dict,
+		Count:  p.count,
+		Nulls:  p.nulls,
+		Min:    p.min,
+		Max:    p.max,
+	}
+}
+
+func (p *PrimitiveWriter) makeDict() []DictEntry {
+	dict := make([]DictEntry, 0, len(p.dict)+p.hasNull)
+	for key, cnt := range p.dict {
+		dict = append(dict, DictEntry{
+			zed.NewValue(p.typ, zcode.Bytes(key)),
+			uint32(cnt),
```

Comment:
This is already a uint32.
```suggestion
			cnt,
```